### ISSUE TITLE
big security risk

### ISSUE THREAD

**Issue by karolherbst** [SkyLandTW/clevo-indicator#11]
This doesn't belong into userspace, never ever. You don't want to run a suid binary in userspace talking with the EC, the proper way to implement this is as a kernel module and user the proper subdev classes and hwmon.

You really should think about what you do here, because that's an unneccessary security risk with _no_ benefit. This belongs into the kernel as a module, period, no discussions about that.

**Comment by GuilloOme** [comment]
Yep, you are right. Do you think you will be able to propose a pull request to change that ?

**Comment by karolherbst** [comment]
@GuilloOme no, I rather work on this project: https://bitbucket.org/lynthium/clevo-xsm-wmi

**Comment by AqD** [comment]
It won't be a pull request. Kernel-based solution requires a module controlling EC and a way of communication (to userspace), plus a GUI client.

I'd rather integrate it with some multi-hardware platform such as lm_sensors or some sort of auto fan-control service, and only have to write the driver. But lm_sensors website is dead and I have no idea where to look for information.

**Comment by karolherbst** [comment]
@AqD https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface

**Comment by patrick-mota** [comment]
@karolherbst thanks for your issue :+1: 

If I understand well it happens because we have two tools in one.
One part for the unity display so the user can change the fan speed from this "gui tool".
The other one is the part that allows to communicate directly to the EC.

So if I change the code to remove all the stuff about the gui and only use the code that allows to interact to the EC (so only command line used to reduce the fan speed) it should be ok ?

Sorry I don't know anything about Linux kernel...

**Comment by karolherbst** [comment]
@patrick-mota nope. In the end you want to have a linux kernel module doing all the EC stuff and then provide access to functionality through the hwmon interface for fan/temperature stuff. There is even a "fan[1-*]_target" for hwmon where you could set the desired fan speed for example.

https://www.mjmwired.net/kernel/Documentation/hwmon/sysfs-interface

In the end there is no reason to do anything clevo specific in userspace, because there are appropiate kernel->userspace interfaces already which would work with every hardware as long as they have an appropiate driver.

**Comment by patrick-mota** [comment]
@karolherbst ok thank you for pointing me to the right direction ! 
I'm gonna try to do it using sysfs-interface.

**Comment by karolherbst** [comment]
well, there is already https://bitbucket.org/lynthium/clevo-xsm-wmi which does all the things. Maybe knowledge could be shared?

**Comment by rafaelgieschke** [comment]
I actually implemented this some time ago in <https://github.com/rafaelgieschke/clevo-xsm-wmi> (well, incidentally based on the work of @karolherbst; and with thanks to @AqD for your helpful work on <https://github.com/SkyLandTW/ClevoECView/blob/master/ec-trace.txt>).

**Comment by michalzxc** [comment]
Did anyone make kernel module for controlling fans at the end?
